openfe/protocols/openmm_afe/base.py

```diff
@@ -0,0 +1,940 @@
+          Identifier for which repeat (aka replica/clone) this Unit is,
+          default 0
+        generation : int, optional
+          Generation counter which keeps track of how many times this repeat
+          has been extended, default 0.
+        """
+        super().__init__(
+            name=name,
+            stateA=stateA,
+            stateB=stateB,
+            settings=settings,
+            alchemical_components=alchemical_components,
+            repeat_id=repeat_id,
+            generation=generation,
+        )
+
+    @staticmethod
+    def _get_alchemical_indices(omm_top: openmm.Topology,
+                                comp_resids: Dict[str, npt.NDArray],
+                                alchem_comps: Dict[str, List[Component]]
+                                ) -> List[int]:
+        """
+        Get a list of atom indices for all the alchemical species
+
+        Parameters
+        ----------
+        omm_top : openmm.Topology
+          Topology of OpenMM System.
+        comp_resids : Dict[str, npt.NDArray]
+          A dictionary of residues for each component in the System.
+        alchem_comps : Dict[str, List[Component]]
+          A dictionary of alchemical components for each end state.
+
+        Return
+        ------
+        atom_ids : List[int]
+          A list of atom indices for the alchemical species
+        """
+
+        # concatenate a list of residue indexes for all alchemical components
+        residxs = np.concatenate(
+            [comp_resids[key] for key in alchem_comps['stateA']]
+        )
+
+        # get the alchemicical residues from the topology
+        alchres = [
+            r for r in omm_top.residues() if r.index in residxs
+        ]
+
+        atom_ids = []
+
+        for res in alchres:
+            atom_ids.extend([at.index for at in res.atoms()])
+
+        return atom_ids
+
+    @staticmethod
+    def _pre_minimize(system: openmm.System,
+                      positions: omm_unit.Quantity) -> npt.NDArray:
+        """
+        Short CPU minization of System to avoid GPU NaNs
+
+        Parameters
+        ----------
+        system : openmm.System
+          An OpenMM System to minimize.
+        positionns : openmm.unit.Quantity
+          Initial positions for the system.
+
+        Returns
+        -------
+        minimized_positions : npt.NDArray
+          Minimized positions
+        """
+        integrator = openmm.VerletIntegrator(0.001)
+        context = openmm.Context(
+            system, integrator,
+            openmm.Platform.getPlatformByName('CPU'),
+        )
+        context.setPositions(positions)
+        # Do a quick 100 steps minimization, usually avoids NaNs
+        openmm.LocalEnergyMinimizer.minimize(
+            context, maxIterations=100
+        )
+        state = context.getState(getPositions=True)
+        minimized_positions = state.getPositions(asNumpy=True)
+        return minimized_positions
+
+    def _prepare(
```

Comment:
For the live code review - I realise that these are taking classmethod modularity to an excess, and the pattern isn't the nicest - the point here is that this is the level of modularity I'm going to need to be able to quickly hotswap various things for derivative methods (ABFEs, SepTop, etc..).